lib/MySQL_HostGroups_Manager.cpp

```diff
@@ -4499,6 +4569,203 @@ void MySQL_HostGroups_Manager::read_only_action(char *hostname, int port, int re
 	free(query);
 }
 
+/**
+ * @brief New implementation of the read_only_action method that does not depend on the admin table.
+ *   The method checks each server in the provided list and adjusts the servers according to their corresponding read_only value.
+ *   If any change has occured, checksum is calculated.
+ *
+ * @param mysql_servers List of servers having hostname, port and read only value.
+ * 
+ */
+void MySQL_HostGroups_Manager::read_only_action_v2(const std::list<std::tuple<std::string,int,int>>& mysql_servers) {
+
+	std::string hostname;
+	int port = -1;
+	int read_only = -1;
+	bool update_mysql_servers_table = false;
+
+	unsigned long long curtime1 = monotonic_time();
+	wrlock();
+	for (const auto& server : mysql_servers) {
+		bool is_writer = false;
+		std::tie(hostname, port, read_only) = server;
+		const std::string& srv_id = hostname + ":::" + std::to_string(port);
+		
+		auto itr = hostgroup_server_mapping.find(srv_id);
+
+		if (itr == hostgroup_server_mapping.end()) {
+			proxy_warning("Server %s:%d not found\n", hostname.c_str(), port);
+			continue;
+		}
+
+		HostGroup_Server_Mapping* host_server_mapping = itr->second.get();
+
+		if (!host_server_mapping)
+			assert(0);
+
+		const std::vector<HostGroup_Server_Mapping::Node>& writer_map = host_server_mapping->get(HostGroup_Server_Mapping::Type::WRITER);
+
+		is_writer = !writer_map.empty();
+
+		if (read_only == 0) {
+			if (is_writer == false) {
+				// the server has read_only=0 (writer), but we can't find any writer, 
+				// so we copy all reader nodes to writer
+				host_server_mapping->copy_if_not_exists(HostGroup_Server_Mapping::Type::WRITER, HostGroup_Server_Mapping::Type::READER);
+
+				if (mysql_thread___monitor_writer_is_also_reader) {
+					// server is also a reader, we copy all nodes from writer to reader (previous reader nodes will be reused)
+					host_server_mapping->copy_if_not_exists(HostGroup_Server_Mapping::Type::READER, HostGroup_Server_Mapping::Type::WRITER);
+				} else {
+					// server can only be a writer
+					host_server_mapping->clear(HostGroup_Server_Mapping::Type::READER);
+				}
+
+				update_mysql_servers_table = true;
```

Comment:
Here we add another message stating that we are going to perform a regeneration of the `mysql_servers_table`:

```
proxy_info("Regenerating table 'mysql_servers' due to actions on server '%s:%d'\n", hostname.c_str(), port);
```

I'm proposing this two messages at beginning and end of the scope because this way they complement the message already issued by `copy_if_not_exists` which is specify which one is the server being created. This way in the logs we would see:

1. Condition for the action that is going to take place.
2. Action that took place for the server (the creation).
3. Consequence of the action, the regeneration of the `mysql_servers_table`.

Reply:
Done